File: platforms/android/keyboard-ui/src/main/java/app/funput/funput/keyboard/ui/localtext/PlainLocalTextComposer.kt
package app.funput.funput.keyboard.ui.localtext

import app.funput.funput.keyboard.model.KeyboardInputMethod

/**
 * A [LocalTextComposing] field that keeps every key exactly as typed: the default for a
 * keyboard view the IME has not given an engine-backed field.
 */
class PlainLocalTextComposer : LocalTextComposing {
    override var text: String = ""
        private set
    override val inputMethod: KeyboardInputMethod? = null

    override fun apply(edit: LocalTextEdit) {
        text = when (edit) {
            is LocalTextEdit.Text -> text + edit.value
            LocalTextEdit.Space -> if (LocalTextSpacing.accepts(text)) "$text " else text
            LocalTextEdit.DeleteBackward -> text.dropLastCodePoint()
        }
    }

    override fun reset() {
        text = ""
    }
}

/** Drops the last code point, so a surrogate pair never leaves half of itself behind. */
fun String.dropLastCodePoint(): String =
    if (isEmpty()) this else substring(0, offsetByCodePoints(length, -1))

File: platforms/ios/Packages/FunputKit/Sources/KeyboardLayout/LocalText/PlainLocalTextComposer.swift
/// A ``LocalTextComposing`` field that keeps every key exactly as typed.
///
/// The default for panels built without an engine — previews, renderer tests — and the
/// behavior an engine-backed field falls back to while Vietnamese is switched off.
@MainActor
public final class PlainLocalTextComposer: LocalTextComposing {
    public private(set) var text = ""

    public init() {}

    public func apply(_ edit: LocalTextEdit) {
        switch edit {
        case let .text(value):
            text.append(contentsOf: value)
        case .space:
            if LocalTextSpacing.accepts(after: text) { text.append(" ") }
        case .deleteBackward:
            if !text.isEmpty { text.removeLast() }
        }
    }

    public func reset() {
        text = ""
    }
}
